src/Components/test/E2ETest/Tests/ThreadingAppTest.cs

```diff
@@ -0,0 +1,130 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.AspNetCore.Components.E2ETest.Infrastructure;
+using Microsoft.AspNetCore.Components.E2ETest.Infrastructure.ServerFixtures;
+using Microsoft.AspNetCore.E2ETesting;
+using OpenQA.Selenium;
+using Xunit.Abstractions;
+
+namespace Microsoft.AspNetCore.Components.E2ETest.Tests;
+
+public class ThreadingAppTest
+    : ServerTestBase<BlazorWasmTestAppFixture<ThreadingApp.Program>>, IDisposable
+{
+    public ThreadingAppTest(
+        BrowserFixture browserFixture,
+        BlazorWasmTestAppFixture<ThreadingApp.Program> serverFixture,
+        ITestOutputHelper output)
+        : base(browserFixture, serverFixture, output)
+    {
+    }
+
+    protected override void InitializeAsyncCore()
+    {
+        Navigate("/", noReload: true);
+        WaitUntilLoaded();
+    }
+
+    [Fact]
+    public void HasTitle()
+    {
+        Assert.Equal("Blazor standalone", Browser.Title);
+    }
+
+    [Fact]
+    public void HasHeading()
+    {
+        Assert.Equal("Hello, world!", Browser.Exists(By.TagName("h1")).Text);
+    }
+
+    [Fact]
+    public void NavMenuHighlightsCurrentLocation()
+    {
+        var activeNavLinksSelector = By.CssSelector(".sidebar a.active");
+        var mainHeaderSelector = By.TagName("h1");
+
+        // Verify we start at home, with the home link highlighted
+        Assert.Equal("Hello, world!", Browser.Exists(mainHeaderSelector).Text);
+        Assert.Collection(Browser.FindElements(activeNavLinksSelector),
+            item => Assert.Equal("Home", item.Text.Trim()));
+
+        // Click on the "counter" link
+        Browser.Exists(By.LinkText("Counter")).Click();
+
+        // Verify we're now on the counter page, with that nav link (only) highlighted
+        Assert.Equal("Counter", Browser.Exists(mainHeaderSelector).Text);
+        Assert.Collection(Browser.FindElements(activeNavLinksSelector),
+            item => Assert.Equal("Counter", item.Text.Trim()));
+
+        // Verify we can navigate back to home too
+        Browser.Exists(By.LinkText("Home")).Click();
+        Assert.Equal("Hello, world!", Browser.Exists(mainHeaderSelector).Text);
+        Assert.Collection(Browser.FindElements(activeNavLinksSelector),
+            item => Assert.Equal("Home", item.Text.Trim()));
+    }
+
+    [Fact]
+    public async Task HasCounterPage()
+    {
+        // Navigate to "Counter"
+        Browser.Exists(By.LinkText("Counter")).Click();
+        Assert.Equal("Counter", Browser.Exists(By.TagName("h1")).Text);
+
+        var countDisplayElement = Browser.Exists(By.CssSelector("h1 + p"));
+
+        // wait and see timer increase
+        await Task.Delay(500);
+        Assert.NotEqual("Current count: 0", countDisplayElement.Text);
```

Comment:
This looks like a race condition that could mean the test is flaky. It's always possible the browser takes longer than 500ms to do this even if we think it should complete in < 100ms.

The normal way we make these tests reliable is to use the wait-assertions, e.g.: something like

```suggestion
        Browser.NotEqual("Current count: 0", () => countDisplayElement.Text);
```

This will wait until the assertion becomes true, or eventually time out with an error after about 10 seconds if it never becomes true.